kratos/solving_strategies/schemes/residual_based_bossak_displacement_scheme.hpp

```diff
@@ -239,75 +242,128 @@ class ResidualBasedBossakDisplacementScheme
+                    r_current_displacement[1] = r_previous_displacement[1] + delta_time * r_previous_velocity[1] + 0.5 * std::pow(delta_time, 2) * r_previous_acceleration[1];
+                }
             }
 
             // For 3D cases
-            if (it_node -> HasDofFor(DISPLACEMENT_Z)) {
-                if (it_node -> IsFixed(ACCELERATION_Z)) {
-                    current_displacement[2] = previous_displacement[2] + delta_time * previous_velocity[2] + std::pow(delta_time, 2) * ( 0.5 * (1.0 -  2.0 * mBossak.beta) * previous_acceleration[2] + mBossak.beta * current_acceleration[2]);
-                } else if (it_node -> IsFixed(VELOCITY_Z)) {
-                    current_displacement[2] = previous_displacement[2] + 0.5 * delta_time * (previous_velocity[2] + current_velocity[2]) + 0.5 * std::pow(delta_time, 2) * previous_acceleration[2] ;
-                } else if (it_node -> IsFixed(DISPLACEMENT_Z) == false) {
-                    current_displacement[2] = previous_displacement[2] + delta_time * previous_velocity[2] + 0.5 * std::pow(delta_time, 2) * previous_acceleration[2];
+            if (accelpos_z > -1) {
+                if (it_node->GetDof(ACCELERATION_Z, accelpos_z).IsFixed()) {
+                    delta_displacement[2] = (r_current_acceleration[2] + mBossak.c3 * r_previous_acceleration[2] +  mBossak.c2 * r_previous_velocity[2])/mBossak.c0;
+                    r_current_displacement[2] =  r_previous_displacement[2] + delta_displacement[2];
+                    predicted_z = true;
+                }
+            }
+            if (velpos_z > -1 && !predicted_z) {
+                if (it_node->GetDof(VELOCITY_Z, velpos_z).IsFixed() && !predicted_z) {
+                    delta_displacement[2] = (r_current_velocity[2] + mBossak.c4 * r_previous_velocity[2] + mBossak.c5 * r_previous_acceleration[2])/mBossak.c1;
+                    r_current_displacement[2] =  r_previous_displacement[2] + delta_displacement[2];
+                    predicted_z = true;
+                }
+            }
+            if (disppos_z > -1 && !predicted_z) {
+                if (!it_node->GetDof(DISPLACEMENT_Z, disppos_z).IsFixed() && !predicted_z) {
+                    r_current_displacement[2] = r_previous_displacement[2] + delta_time * r_previous_velocity[2] + 0.5 * std::pow(delta_time, 2) * r_previous_acceleration[2];
                 }
             }
-
 
             // Updating time derivatives ::: Please note that displacements and its time derivatives can not be consistently fixed separately
-            noalias(delta_displacement) = current_displacement - previous_displacement;
+            noalias(delta_displacement) = r_current_displacement - r_previous_displacement;
 
-            UpdateVelocity     (current_velocity,     delta_displacement, previous_velocity, previous_acceleration);
+            UpdateVelocity(it_node, r_current_velocity, delta_displacement, r_previous_velocity, r_previous_acceleration);
```

Comment:
you do not need to pass the It_node if you don't check for fixity! (and pelase don't as i am commenting before).

So these functions can remain unmodified

Reply:
Ok